hrms/overrides/expense_claim.py

```diff
@@ -0,0 +1,323 @@
+            )
+
+            amount_via_payment_entry = frappe.db.get_value(
+                "Payment Entry Reference",
+                {"reference_name": self.name, "docstatus": 1},
+                "sum(allocated_amount)",
+            )
+
+            return flt(amount_via_jv) + flt(amount_via_payment_entry)
+
+
+@frappe.whitelist(allow_guest=True)
+@handle_exceptions_with_readable_message
+def make_gl_entries(doc_name):
+    import pdb
+
+    pdb.set_trace()
+    expense_claim_doc = frappe.get_doc("Expense Claim", doc_name)
+    expense_claim_doc.on_submit()
+
+
+@frappe.whitelist()
+@handle_exceptions_with_readable_message
+def next_state(doc_name):
+    expense_claim_doc = frappe.get_doc("Expense Claim", doc_name)
+
+    if expense_claim_doc.status == ExpenseClaimConstants.DRAFT:
+        if frappe.session.user != expense_claim_doc.owner:
+            frappe.throw(_("Only Owner can approve in draft state"))
+        CustomTask.create_link_doc_task_if_all_parents_closed(
+            None,
+            TaskType.APPROVE_EXPENSE_CLAIM,
+            "Expense Claim",
+            expense_claim_doc.name,
+            None,
+            None,
+            [expense_claim_doc.approver_1],
+        )
+        expense_claim_doc.status = ExpenseClaimConstants.PENDING_APPROVAL
+        expense_claim_doc.save()
+    elif expense_claim_doc.status == ExpenseClaimConstants.PENDING_APPROVAL:
+        if (
+            frappe.session.user != expense_claim_doc.approver_1
+            and frappe.session.user != "Administrator"
+        ):
+            frappe.throw(_("Only the added approver can approve the document"))
+        expense_claim_doc.status = ExpenseClaimConstants.PENDING_APPROVAL_BY_ADMIN
+        CustomTask.mark_tasks_as_completed(
+            "Expense Claim", expense_claim_doc.name, TaskType.APPROVE_EXPENSE_CLAIM
+        )
+        expense_claim_doc.save()
+        expense_claim_doc.submit()
+    elif expense_claim_doc.status == ExpenseClaimConstants.PENDING_APPROVAL_BY_ADMIN:
+        if not user_has_role(frappe.session.user, RoleEnum.OFFICE_ADMIN_ROLE.value):
+            frappe.throw(_("Only Admin approve the document"))
+
+        expense_claim_doc.status = ExpenseClaimConstants.PENDING_APPROVAL_BY_HR
+        expense_claim_doc.save()
+    elif expense_claim_doc.status == ExpenseClaimConstants.PENDING_APPROVAL_BY_HR:
+        if not user_has_role(frappe.session.user, RoleEnum.HR_ROLE.value):
+            frappe.throw(_("Only user having HR User can approve the document"))
+        if expense_claim_doc.grand_total == 0:
+            expense_claim_doc.status = ExpenseClaimConstants.PAID
+        else:
+            expense_claim_doc.status = ExpenseClaimConstants.PENDING_PAYMENT
+        expense_claim_doc.save()
+
+
+@frappe.whitelist()
+@handle_exceptions_with_readable_message
+def cancel_doc(doc_name):
+    expense_claim_doc = frappe.get_doc("Expense Claim", doc_name)
```

Comment:
same as employee_advance, check for statuses and roles

Reply:
Done